backend/src/main/java/de/codexbella/ContentService.java

```diff
@@ -1,46 +1,76 @@
       this.restTemplate = restTemplate;
       this.apiKey = apiKey;
+      this.showRepository = showRepository;
+      this.contentMapper = contentMapper;
    }
 
-   public List<ShowSearchData> searchForShows(String language, String searchTerm) {
+   public List<ShowSearchData> searchForShows(String language, String searchTerm, String username) {
       String response = restTemplate.getForObject(
-            "https://api.themoviedb.org/3/search/tv?api_key="+apiKey+"&language="+language+"&query="+searchTerm, String.class);
+            "https://api.themoviedb.org/3/search/tv?api_key="+apiKey+"&language="+language+"&query="
+                  +searchTerm, String.class);
       SearchResultShows results = new Gson().fromJson(response, SearchResultShows.class);
       Stream<ShowSearchData> resultListStream = results.getShows().stream();
       if (results.getNumberOfPages()>1) {
          for (int i = 2; i <= results.getNumberOfPages(); i++) {
             String toAdd = restTemplate.getForObject(
-                  "https://api.themoviedb.org/3/search/tv?api_key="+apiKey+"&language="+language+"&query="+searchTerm+"&page="+i, String.class);
+                  "https://api.themoviedb.org/3/search/tv?api_key="+apiKey+"&language="+language+"&query="
+                        +searchTerm+"&page="+i, String.class);
             SearchResultShows resultsAdditional = new Gson().fromJson(toAdd, SearchResultShows.class);
             Stream<ShowSearchData> resultListAdditionalStream = resultsAdditional.getShows().stream();
             resultListStream = Stream.concat(resultListStream, resultListAdditionalStream);
          }
       }
-      return resultListStream.distinct().toList();
+      List<ShowSearchData> result = resultListStream.distinct().toList();
+      for (int i = 0; i < result.size(); i++) {
+         if (showRepository.findByApiIdAndUsername(result.get(i).getApiId(), username).isPresent()) {
+            result.get(i).setLiked(true);
+         }
+      }
+      return result;
    }
 
-   public Show addShow(String language, int id) {
-      String response = restTemplate.getForObject(
-            "https://api.themoviedb.org/3/tv/"+id+"?api_key="+apiKey+"&language="+language, String.class);
-      return new Gson().fromJson(response, Show.class);
+   public void saveShow(String language, int apiId, String username) throws IllegalArgumentException {
+      Optional<Show> showOptional = showRepository.findByApiIdAndUsername(apiId, username);
+      if (showOptional.isEmpty()) {
+         String response = restTemplate.getForObject(
+               "https://api.themoviedb.org/3/tv/" + apiId + "?api_key=" + apiKey + "&language=" + language, String.class);
+         ShowApi showApi = new Gson().fromJson(response, ShowApi.class);
+         Show show = contentMapper.toShow(showApi);
+         show.setUsername(username);
+         for (int i = 0; i < show.getSeasons().size(); i++) {
```

Comment:
This is a good example for the `flatMap` method.
```java
show.getSeasons().stream()
    .flatMap(season -> {
        season.setUsername(username);
        return season.getEpisodes().stream();
    })
    .forEach(episode -> episode.setUsername(username);
```
If you don't want to use the stream, I would still recommend the `foreach`-loop, because you don't know what list implementation is returned by Gson. If `LinkList`s are returned the index loop would be very slow (https://neuefische-students.slack.com/archives/C0302PPVC2J/p1644430897499769).